### DISCUSSION TITLE
How should we design the CitSci Package and Notebooks together?

### DISCUSSION THREAD

**Discussion by bnord** [lsst-epo/citizen-science-notebooks#44] [Ideas]
@jsv1206 @beckynevin @clareh @ericdrosas87 

# Goals
1. allow simple user interface for those who don't want to do anything out of the norm
2. allow people to make and implement their own stuff
3. Package the code for reproducibility and readability

# Considerations and Questions
1. Who should maintain the notebooks?
2. Who should maintain the package?
3. Are the notebooks and the package truly different?

**Comment by bnord** [comment]
Claim: Well-maintained packaged code is equivalent (in terms of openness) to a deeply explained, exposed notebook.

**Reply by bnord** [reply]
We can write tutorials about packaged code that show how scripts work, and how they can be changed by a user. Rubin scientists should be acclimated to this as a standard, given the rest of the Rubin coding infrastructure.

**Reply by ericdrosas87** [reply]
> We can write tutorials about packaged code that show how scripts work, and how they can be changed by a user.

I'm not sure if this is what you mean, but[ it's highly discouraged to modify any installed third party package](https://stackoverflow.com/questions/23075397/python-how-to-edit-an-installed-package#:~:text=You%20should%20never%20edit%20an,%2Fsite_packages%2F...'&text=Put%20your%20changes%20in%20a,pip%20to%20install%20it%20explicitly.) in the software development world.

If you mean that RSP users will be able to fork the code that is packed up as PyPI package from the official Github repo, then it becomes a matter of: what is the level of effort for supporting those RSP users who fork and modify their version of the code and do we have the resources to do achieve that level of support.

**Reply by bnord** [reply]
I just mean the packages we make, not other folks's stuff.

**Reply by bnord** [reply]
It looks to me that if we leave all of our low-level-ish code exposed via the notebooks, then we're inviting requests of support for that anyway. What do you think?

I envision that if we script and package most of that code then we give a boon of simplicity to most users, while still providing open-source options to super-users, which I'm guessing will be a much smaller group. Would that work?

I imagine we could establish some contribution and support policies that make it clear what level of support we're willing to provide for code updates. 

I'm still relatively new to software development from the inside of the process, so I'm making some semi-educated guesses here.

**Reply by ericdrosas87** [reply]
>I just mean the packages we make, not other folks's stuff.

Ah okay, I think I was misinterpreting your comment. Just to be doubly sure that we're on the same page:

* People directly involved with Rubin (all us) can modify and publish to PyPI as we see fit
* External people should not be modifying the package we develop after they install it via `pip install`, if they want to make any changes they should be directed to fork the repository the package we develop is based on

**Reply by ericdrosas87** [reply]
>It looks to me that if we leave all of our low-level-ish code exposed via the notebooks, then we're inviting requests of support for that anyway. What do you think?

I can see that - although we can disable editing on the cells of the SDK notebook too. Not that the RSP user couldn't re-enable the editing of the notebook if they know what they're doing.

**Reply by beckynevin** [reply]
Okay so I'm hearing that the pypi version that's pip installable would be totally controlled by us, but experienced users could fork the repo and suggest new features. Would we need to offer support to the people who want to raise issues on the repo? And at what point would changes via fork be incorporated into the pypi version. I guess what I'm wondering is how will the pypi and the repo compare?

**Reply by beckynevin** [reply]
Lol and how much work is it to support people who are raising issues/ suggesting changes? Or is there a way to not allow people to raise issues (that we "the developers" have to address) and instead only allow them to suggest changes?

**Reply by bnord** [reply]
I think we're allowed to say what kinds of support we're prepared to provide. We might want to consult Rubin guidelines to check if there's something we can learn from or align with.

**Reply by beckynevin** [reply]
Yeah, let's check docu, but it feels okay if we can just make it a policy from the beginning that "we won't create features for you but if you want to fork and submit a PR we will consider" something along those lines...

**Reply by ericdrosas87** [reply]
> I guess what I'm wondering is how will the pypi and the repo compare?

I want to say that PyPI is essentially an index that points to Github repositories upon `pip install`. See: the [Rubin DM Butler package page](https://pypi.org/project/lsst-daf-butler/) for reference. You'll notice on the left-hand side of the page you'll see a `Project links` section with a `Homepage` link that just links to the Github repository. So conceivably an external (not us) user of the citSci package would find the repo that way if they were interested in forking it to make their own changes.

It is a common practice in the software development world to allow for external users to open up Github Issues and I think we should offer that level of support, if possible, as external users could then report not just feature enhancements, but also bug reports. It's also not uncommon for issues to remain open for quite awhile (years even) if they are feature requests, or for feature requests to be closed out for reasons such as "not implementing in the near future"/"we have no plans to implement this feature" - so developer-type folks _should_ be used to seeing those types of responses.

**Comment by bnord** [comment]
Proposal: we prioritize the packaged modules and then add increasing levels of explanations of that code to let users make what they need.

**Reply by bnord** [reply]
This would allow us to start small with documentation, and have something very clear and clean for PCW or earlier milestones.
